src/adapters/bidderFactory.js

```diff
@@ -0,0 +1,240 @@
+              {
+                method: GET,
+                withCredentials: true
+              }
+            );
+            break;
+          case POST:
+            ajax(
+              request.endpoint,
+              {
+                success: onSuccess,
+                error: onFailure
+              },
+              typeof request.data === 'string' ? request.data : JSON.stringify(request.data),
+              {
+                method: POST,
+                contentType: 'text/plain',
+                withCredentials: true
+              }
+            );
+            break;
+          default:
+            logWarn(`Skipping invalid request from ${spec.code}. Request type ${request.type} must be GET or POST`);
+            onResponse();
+        }
+      }
+
+      // If the server responds successfully, use the adapter code to unpack the Bids from it.
+      // If the adapter code fails, no bids should be added. After all the bids have been added, make
+      // sure to call the `onResponse` function so that we're one step closer to calling fillNoBids().
+      function onSuccess(response) {
+        const bidRequestMap = {};
+        bidRequests.forEach(bid => {
+          bidRequestMap[bid.bidId] = bid;
+        });
+        function addBidUsingRequestMap(bid) {
+          const bidRequest = bidRequestMap[bid.requestId];
+          if (bidRequest) {
+            const prebidBid = Object.assign(bidfactory.createBid(STATUS.GOOD, bidRequest), bid);
+            addBidWithCode(bidRequest.placementCode, prebidBid);
+          } else {
+            logWarn(`Bidder ${spec.code} made bid for unknown request ID: ${bid.requestId}. Ignoring.`);
+          }
+        }
+
+        let bids;
+        try {
+          bids = spec.interpretResponse(response);
+        } catch (err) {
+          logError(`Bidder ${spec.code} failed to interpret the server's response. Continuing without bids`, null, err);
+          onResponse();
+          return;
+        }
+
+        if (bids) {
+          if (bids.forEach) {
+            bids.forEach(addBidUsingRequestMap);
+          } else {
+            addBidUsingRequestMap(bids);
+          }
+        }
+        onResponse();
+      }
+
+      // If the server responds with an error, there's not much we can do. Log it, and make sure to
+      // call onResponse() so that we're one step closer to calling fillNoBids().
+      function onFailure(err) {
+        logError(`Server call for ${spec.code} failed: ${err}. Continuing without bids.`);
+        onResponse();
+      }
+    }
+  });
+
+  function filterAndWarn(bid) {
+    if (!spec.areParamsValid(bid.params)) {
+      logWarn(`Invalid bid sent to bidder ${spec.code}: ${JSON.stringify(bid)}`);
+      return false;
+    }
+    return true;
+  }
+
+  function newEmptyBid() {
+    const bid = bidfactory.createBid(STATUS.NO_BID);
+    bid.code = spec.code;
+    bid.bidderCode = spec.code;
+    return bid;
+  }
+}
+
+/**
+ * @typedef {object} ServerRequest
```

Comment:
I've been thinking about this... but I'm not so convinced.

First: `instanceof` is a pretty flimsy check in javascript. For example:

```
var serverRequest = new ServerRequest(...);
delete serverRequest.type

serverRequest.type // Is undefined
serverRequest instanceof Request // Is true
```

Second: it sounds like all the complex validation code has just moved into the constructor. Rather than `if (!serverRequest.type)`, the constructor has to do `if (!type)`. I can see some advantage in having the validation code centralized... but if that's the goal, I'd rather define it an `isRequest(obj)` function somewhere.

Third: constructors are a bad pattern. To be honest, I don't expect the drawbacks to affect this area much... but I do want to discourage them from spreading around the codebase.  There's some good discussion on them in (2) and (3) [here](https://medium.com/javascript-scene/javascript-factory-functions-vs-constructor-functions-vs-classes-2f22ceddf33e)

Third: I'm open to discussion here... but I was kind of hoping we can _avoid_ validating the data at runtime. React did the profiling and found that data validation was such a performance drain that they made separate "dev" and "prod" versions of their library. The "prod" one skips all property validation in your Components.

----------------
Data validation is an open question as far as I'm concerned, though.

I left it out here because I was thinking of how React totally disabled all runtime property validation in the `prod` version of their library because it was found to be a notable performance drain.

I _think_ it should be unnecessary so long as the unit tests check for it. I have a followup PR I hope to make for "unified adapter tests"--assuming management schedules the time for me to do it--which would do the validation at test time, rather than runtime.

Reply:
It's true, there are ways to subvert `instanceof` checks.  The reason I like them is because, as you said, type validation is expensive and `instanceof` checks are not.  Using a constructor function to define a type allows you to incur a one-time hit for the validation of an object (and then use `instanceof` thereafter), as opposed to `isRequest(obj)` which is a many-time hit for each time that object is used.  That's the problem React encountered as they had to validate propTypes each type a component instance was used.

Either way, this isn't a make or break issue for me, it's just something I think of every time I see `if (obj.property ...`.  Also, if we did do this we'd have to expose the constructor functions on the public API as well so external adapters could pass the `instanceof` checks, which isn't terrible but it's more of a hassle than just using typedef.

Reply:
yeah... I do understand where you're coming from, but I think I'd prefer to hold off on this for now.